test/e2e_node/containers_lifecycle_test.go

```diff
@@ -0,0 +1,605 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2enode
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+)
+
+type execCommand struct {
+	ExitCode int
+	Delay    int
+}
+
+func ExecCommand(name string, c execCommand) []string {
+	var cmd bytes.Buffer
+	// all outputs are in the format of:
+	// time-since-boot timestamp container-name message
+
+	// The busybox time command doesn't support sub-second display. uptime displays in hundredths of a second, so we
+	// include both and use time since boot for relative ordering
+	timeCmd := "`date +%s` `cat /proc/uptime | awk '{print $1}'`"
+	containerLog := fmt.Sprintf("/persistent/%s.log", name)
+
+	fmt.Fprintf(&cmd, "touch %s; ", containerLog)
+	fmt.Fprintf(&cmd, "cat %s >> /dev/termination-log; ", containerLog)
+
+	fmt.Fprintf(&cmd, "echo %s '%s Starting' | tee -a %s >> /dev/termination-log; ", timeCmd, name, containerLog)
+	fmt.Fprintf(&cmd, "echo %s '%s Delaying %d' | tee -a %s >> /dev/termination-log; ", timeCmd, name, c.Delay, containerLog)
+	if c.Delay != 0 {
+		fmt.Fprintf(&cmd, "sleep %d; ", c.Delay)
+	}
+	fmt.Fprintf(&cmd, "echo %s '%s Exiting'  | tee -a %s >> /dev/termination-log; ", timeCmd, name, containerLog)
+	fmt.Fprintf(&cmd, "exit %d", c.ExitCode)
+	return []string{"sh", "-c", cmd.String()}
+}
+
+// WaitForPodContainerRestartCount waits for the given Pod container to achieve at least a given restartCount
+func WaitForPodContainerRestartCount(ctx context.Context, c clientset.Interface, namespace, podName string, containerIndex int, desiredRestartCount int32, timeout time.Duration) error {
```

Comment:
@tzneal @andrewsykim  moved method here temporarily so I can self-approve this PR and rebase the sidecar container branch

Reply:
@tzneal can you please re-lgtm?